### ISSUE TITLE
transmission-create memory leak

### ISSUE THREAD

**Issue by greyltc** [transmission/transmission#5139]
### What is the issue?

Memory usage of `transmission-create directoy_name` seems to grow forever. `directory_name` is ~11GB and has ~30k files. I've watched memory usage climb to 30 GB in 5 min before killing the process. Arch Linux.

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.1

**Comment by ckerr** [comment]
@greyltc could you please run `transmission-create` in `heaptrack` and attach the resulting datafile to this ticket?

Literally `heaptrack transmission-create ...` :smile_cat:

**Comment by greyltc** [comment]
I've attached 5 or 6 seconds of that. If that's not enough, let me know!
[heaptrack.transmission-create.32557.zst.zip](https://github.com/transmission/transmission/files/10888925/heaptrack.transmission-create.32557.zst.zip)

Delete the .zip from the file name here before use, I just added that to fool github's dumb upload filter.

**Comment by greyltc** [comment]
By the way, I've also built the latest transmission from git and the issue is there too.

**Comment by ckerr** [comment]
@greyltc I'm not able to read that:

```sh
$ wget --quiet https://github.com/transmission/transmission/files/10888925/heaptrack.transmission-create.32557.zst.zip
$ mv heaptrack.transmission-create.32557.zst.zip heaptrack.transmission-create.32557.zst
$ zstd -d heaptrack.transmission-create.32557.zst
ion-create.32557.zst : 0 MB...     ion-create.32557.zst : Read error (39) : premature end 
```

**Comment by greyltc** [comment]
hm. works for me:
```
wget --quiet https://github.com/transmission/transmission/files/10888925/heaptrack.transmission-create.32557.zst.zip -O heaptrack.zst
heaptrack heaptrack.zst
```
![image](https://user-images.githubusercontent.com/16778174/222916734-5dc78853-1178-4cec-a441-bc4801ca1ca0.png)

**Comment by ckerr** [comment]
Hm, might've been PEBCAK issue; your recipe wfm.

All I'm seeing are <unresolved functions>, is this on a build whose debugging symbols have been stripped out?

For testing, at least, might want to build with `-DCMAKE_BUILD_TYPE=Debug`

**Comment by greyltc** [comment]
cool, let me see if i can fix that up

**Comment by greyltc** [comment]
Hope I've got what you need this time!
[heaptrack.transmission-create.12514.zst.zip](https://github.com/transmission/transmission/files/10889175/heaptrack.transmission-create.12514.zst.zip)

**Comment by ckerr** [comment]
Huh. That thing says it's got 2.2 mil st::string allocations in TorrentFile (a struct which owns two strings) while walking a directory tree that's got 30k files in it :confused: 

It looks like maybe the directory walk is getting stuck somehow. Are there any symbolic links in the subdirectory that might confuse a directory walker into looping?

Maybe try adding this a tracer statement to look for any patterns / repetitions:

```diff
diff --git a/libtransmission/makemeta.cc b/libtransmission/makemeta.cc
index 108b78f1a..d99a32063 100644
--- a/libtransmission/makemeta.cc
+++ b/libtransmission/makemeta.cc
@@ -102,6 +102,8 @@ void walkTree(std::string_view const top, std::string_view const subpath, std::s
                     continue;
                 }
 
+                fmt::print("{:s}:{:d} dir '{:s}' name '{:s}'\n", __FILE__, __LINE__, path, name);
+
                 if (!std::empty(subpath))
                 {
                     walkTree(top, tr_pathbuf{ subpath, '/', name }, files);
```

**Comment by greyltc** [comment]
Yes, there are symlinks. Maybe that's what's confusing the walker? Let me see if I can make something reproducible for you.

**Comment by greyltc** [comment]
My archive contained a special build of proton. I bet something in proton's tree is causing issues for the walker.
I think this should allow you to recreate the issue on your end:
```
wget https://github.com/GloriousEggroll/proton-ge-custom/releases/download/4.14-GE-1/Proton-4.14-GE-1.tar.gz
tar -zxf Proton-4.14-GE-1.tar.gz
transmission-create Proton-4.14-GE-1
```

**Comment by ckerr** [comment]
```
./dist/share/default_pfx/dosdevices:
total 0
lrwxrwxrwx 1 charles charles 10 Aug 20  2019 c: -> ../drive_c
lrwxrwxrwx 1 charles charles  1 Aug 20  2019 z: -> /
```

Oh I see at least one problem -- you're trying to create a torrent that links your root directory, so you're trying to build a torrent of your entire computer.

This doesn't smell like a Transmission bug :smile_cat:

**Comment by greyltc** [comment]
ooooooh. I had no idea a link to `/` was buried in the file structure! How about a new `transmission-create --symlinks` option where any symbolic links the walker finds are not dereferenced/traversed into and instead treated as actual symbolic link files?

Or better yet, maybe the walker shouldn't follow them by default and the option should be `--follow-symlinks`. In any case, it seems like it would be a great feature to add the ability to include actual symbolic link files somehow. Maybe the use case of torrenting an unpacked https://github.com/GloriousEggroll/proton-ge-custom/releases/download/4.14-GE-1/Proton-4.14-GE-1.tar.gz archive is a valid one?

Maybe I'm weird (okay, not maybe :-P), but I am surprised they're walked into by default.

**Comment by ckerr** [comment]
> How about a new transmission-create --symlinks option where any symbolic links the walker finds are not dereferenced/traversed into and instead treated as actual symbolic link files?

No, probably not. Any additional flag has to be implemented in transmission-create + the macOS, GTK, and Qt clients... ugh. At its root the problem here is exactly what you said: "I had no idea a link to / was buried in the file structure" so maybe a warning (or literally just erroring out) would be better.

symlinks in a torrent don't make a lot of sense so now that I think about it erroring out might be The Right Thing to do.

**Comment by greyltc** [comment]
> symlinks in a torrent don't make a lot of sense

I hope you're open to reconsidering that!

**Comment by ckerr** [comment]
> > symlinks in a torrent don't make a lot of sense
> 
> I hope you're open to reconsidering that!

Well, persuade me. What's the use case for them here?

**Comment by greyltc** [comment]
I'm probably biased, but torrenting an unpacked special build of https://github.com/GloriousEggroll/proton-ge-custom/releases/download/4.14-GE-1/Proton-4.14-GE-1.tar.gz feels like a totally legit torrent use case.

Symbolic links are used heavily between the .so files typically found in `/lib/` in Linux. I hope that alone is a good enough reason for the feature to exist!

**Comment by valcomm** [comment]
I don't think there should be any difference between regular file/regular dir/symlink/hardlink.
At most error out on found cycles.
Behavior should be explicitly stated in docs/help and leave it up to the user how to proceed the case (maybe helping `find -type l -links` command to find symlinks/hardlinks so user would know)
I use creating torrents from symlinks very often.  It's easier to symlink many individual files from different locations in some directory, create .torrent and seed that directory.  If symlinks were ignored, I'd have to copy all those files.  And each case is like tens of GiB.
`transmission-create` is an user utility, not a file system utility.  No need to bring in that functionality and complexity in it by my opinion.

**Comment by greyltc** [comment]
@valcomm I think it's arguably less complicated and error prone to treat symbolic links as files than to walk into them (you can walk into loops as we've found here). If the symlink behavior did change, then maybe your workflow could be to use hard links instead?

**Comment by greyltc** [comment]
By the way, I actually think the initial title of this issue should stand. `transmission-create` should probably not consume endless memory if it encounters a symlink that points behind itself. I wonder if a simple symbolic link to `..` will do this...
```bash
mkdir looptest
ln -s .. looptest/looper
transmission-create looptest
rm -rf /  # don't worry, your system is fine :-)
```
yes it will!

**Comment by valcomm** [comment]
> @valcomm I think it's arguably less complicated and error prone to treat symbolic links as files than to walk into them (you can walk into loops as we've found here). 

That's why I mentioned that finding loops might be a good idea.  If, again, it is on `transmission-create` and not on user.  By my opinion if user crated a loop - that's his/her responsibility.  Some utilities add command line have switches for such cases, some don't.


> If the symlink behavior did change, then maybe your workflow could be to use hard links instead?

Could you remind me how to create hardlinks across different partitions, please? ;-)

> Or another solution: `transmission-create` gets a new feature to accept lists of inputs instead of just one!

`transmission-create` needs files and relative paths to them.  This could be a way - feed output of `find` to `transmission-create`.  
Though, again, this is a solving of a complicated user problem.  The only critical part is loops because it leads to a noticeable failure.  And even this might be detected/worked around by other utilities.

**Comment by deliciouslytyped** [comment]
Symbolic links seem to be specified in BEP 47 (https://www.bittorrent.org/beps/bep_0047.html) which is implemented by some torrent libraries.

I'm currently looking for a (cli) client I can create such torrents with.
